common/raw_hashtable_metadata_group.h

```diff
@@ -0,0 +1,930 @@
+      __builtin_assume(mask_ != 0);
+      index_ = BitIndexT(mask_).index();
+      return index_;
+    }
+
+    template <typename T>
+    auto index_ptr(T* pointer) -> T* {
+      return BitIndexT(mask_).index_ptr(pointer);
+    }
+
+    auto operator++() -> Iterator& {
+      CARBON_DCHECK(mask_ != 0) << "Must not increment past the end!";
+      __builtin_assume(mask_ != 0);
+      mask_ &= (mask_ - 1);
+      return *this;
+    }
+
+   private:
+    ssize_t index_;
+    MaskT mask_ = 0;
+  };
+
+  BitIndexRange() = default;
+  explicit BitIndexRange(MaskT mask) : mask_(mask) {}
+
+  explicit operator bool() const { return !empty(); }
+  auto empty() const -> bool { return BitIndexT(mask_).empty(); }
+
+  auto begin() const -> Iterator { return Iterator(mask_); }
+  auto end() const -> Iterator { return Iterator(); }
+
+  template <int N>
+  auto Test() const -> bool {
+    return mask_ & (static_cast<MaskT>(1) << N);
+  }
+
+  friend auto operator==(BitIndexRange lhs, BitIndexRange rhs) -> bool {
+    return lhs.mask_ == rhs.mask_;
+  }
+
+  auto Print(llvm::raw_ostream& out) const -> void {
+    out << llvm::formatv("{0:x}", mask_);
+  }
+
+  explicit operator MaskT() const { return mask_; }
+  explicit operator BitIndexT() const { return BitIndexT(mask_); }
+
+ private:
+  MaskT mask_ = 0;
+};
+
+// A group of metadata bytes that can be manipulated together.
+//
+// The metadata bytes used Carbon's hashtable implementation are designed to
+// support manipulating as groups, either using architecture specific SIMD code
+// sequences or using portable SIMD-in-an-integer-register code sequences. These
+// operations are unusually performance sensitive and in sometimes surprising
+// ways. The implementations here are crafted specifically to optimize the
+// particular usages in Carbon's hashtable and should not be expected to be
+// reusable in any other context.
+//
+// Throughout the functions operating on this type we use the following pattern
+// to have a fallback portable implementation that can be directly used in the
+// absence of a SIMD implementation, but to have the *exact* code for that
+// portable implementation also used to check that any SIMD implementation
+// produces the same result as the portable one. This structure ensures we don't
+// have any un-compiled or un-tested path through the portable code even on
+// platforms where we use SIMD as we expect to practically only test on
+// platforms with a SIMD implementation and so be at a high risk of bit-rot.
+//
+// ```cpp
+// auto Operation(...) -> ... {
+//   ... portable_result;
+//   if constexpr (!UseSIMD || DebugChecks) {
+//     portable_result = PortableCode(...);
+//     if (!UseSIMD) {
+//       return portable_result;
+//     }
+//   }
+//   ... result;
+// #if CARBON_NEON_SIMD_SUPPORT
+//   result = NeonCode(...);
+// #elif CARBON_X86_SIMD_SUPPORT
+//   result = X86Code(...);
+// #else
+//   static_assert(!UseSIMD, "Unimplemented SIMD operation");
+// #endif
```

Comment:
Code below seems to follow a different pattern:
```suggestion
//   ... result = SIMDCode(...);
```